common/src/services/keyConnector.service.ts

```diff
@@ -0,0 +1,86 @@
+import { KeyConnectorService as KeyConnectorServiceAbstraction } from '../abstractions/keyConnector.service';
+import { LogService } from '../abstractions/log.service';
+import { StorageService } from '../abstractions/storage.service';
+import { UserService } from '../abstractions/user.service';
+
+import { OrganizationUserType } from '../enums/organizationUserType';
+
+import { Utils } from '../misc/utils';
+
+import { SymmetricCryptoKey } from '../models/domain/symmetricCryptoKey';
+
+import { KeyConnectorUserKeyRequest } from '../models/request/keyConnectorUserKeyRequest';
+
+const Keys = {
+    usesKeyConnector: 'usesKeyConnector',
+};
+
+export class KeyConnectorService implements KeyConnectorServiceAbstraction {
+    private usesKeyConnector: boolean = false;
+
+    constructor(private storageService: StorageService, private userService: UserService,
+        private cryptoService: CryptoService, private apiService: ApiService,
+        private environmentService: EnvironmentService, private logService: LogService) { }
+
+    setUsesKeyConnector(usesKeyConnector: boolean) {
+        this.usesKeyConnector = usesKeyConnector;
+        return this.storageService.save(Keys.usesKeyConnector, usesKeyConnector);
+    }
+
+    async getUsesKeyConnector(): Promise<boolean> {
+        return this.usesKeyConnector ??= await this.storageService.get<boolean>(Keys.usesKeyConnector);
+    }
+
+    async userNeedsMigration() {
+        const requiredByOrganization = await this.getManagingOrganization() != null;
+        const userIsNotUsingKeyConnector = !await this.getUsesKeyConnector();
+
+        return requiredByOrganization && userIsNotUsingKeyConnector;
+    }
+
+    async migrateUser() {
+        const organization = await this.getManagingOrganization();
+        const key = await this.cryptoService.getKey();
+
+        try {
+            const keyConnectorRequest = new KeyConnectorUserKeyRequest(key.encKeyB64);
+            await this.apiService.postUserKeyToKeyConnector(organization.keyConnectorUrl, keyConnectorRequest);
+        } catch (e) {
+            throw new Error('Unable to reach key connector');
+        }
+
+        await this.apiService.postConvertToKeyConnector();
+    }
+
+    async getAndSetKey(url?: string) {
+        if (url == null) {
+            url = this.environmentService.getKeyConnectorUrl();
+        }
+
+        if (url == null) {
+            throw new Error('No Key Connector URL found.');
+        }
+
+        try {
+            const userKeyResponse = await this.apiService.getUserKeyFromKeyConnector(url);
+            const keyArr = Utils.fromB64ToArray(userKeyResponse.key);
+            const k = new SymmetricCryptoKey(keyArr);
+            await this.cryptoService.setKey(k);
+        } catch (e) {
+            this.logService.error(e);
+            throw new Error('Unable to reach key connector');
+        }
+    }
+
+    async getManagingOrganization() {
```

Comment:
It's probably a bit redundant to check if the user is a `providerUser` since those always gets the role owner. But should we ever change that logic it's good to have the check.

Reply:
I didn't realise that, it's confusing to me that the `Owner` role does not just equal `OrganizationUserType.Owner`. But if that's the pattern that we're following, then I think we should follow it and not introduce the redundant check.

Reply:
Wait, `organization.type` returns `OrganizationUserType`, but `get isOwner()` returns `OrganizationUserStatusType.Owner || isProviderUser`? Seems bad to me; I will still refactor to use the getter though